# VALIDATION AND CROSS VALIDATION

<b> Measuring the model success again on the data used to develop the model may not give us reliable results. We have already found the model using this data. The success of a model is measured by the success of its predictions using data it has never seen before. So where do we find this new data?
The general practice is to divide the data we have into two and test the model we developed with the second one while developing the model with the first one. These two models are called the learning set (traning set) and the validation set (test set). Although it is not a general rule, 70% or 80% of the main data is 20% or 30% of the learning cluster is called the validation cluster. For a healthy model development, the learning and validation clusters must be randomly divided.</b>

<b>Python sklearn module has train_test_split function which is used to split data into learning and validation clusters.</b>

In [11]:
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import pandas as pd

bostondf = pd.read_csv('Boston.csv')
lstat = bostondf[['lstat']].values
medv = bostondf['medv'].values

X_train , X_test ,Y_train , Y_test =train_test_split(lstat,medv,test_size=0.2,random_state=123)

<b>X_train and Y_train represent the dependent and non-dependent variables that will be used to develop the model, and X_test and Y_test to test the model. Test_size specifies the ratio of the data to be reserved for validation (test) and takes the value 0.25 by default. Random_state is used to generate random numbers.</b>

<b>For example, if the data set we have includes the data for the years 2000 and 2010, we want the data for all years to be distributed to the development and test sets at the same rate (for example, 80% and 20%). In this case, we need to specify the column showing the years instead of y.</b>

In [14]:
model = LinearRegression()
model.fit(X_train,Y_train)
Y_estimate = model.predict(X_test)

rmse = np.sqrt(mean_squared_error(Y_test,Y_estimate))
rmse

6.494118435984956

In [15]:
RSS = sum((Y_test - Y_estimate)**2)
TSS = sum((Y_test - np.mean(Y_test))**2)
R2 = 1 - RSS/TSS
R2

0.4902618098232453

In [16]:
n = len(Y_test)
d = X_test.shape[1]
adj_R2 = 1 - (RSS/(n-d-1))/(TSS / (n-1))
adj_R2

0.48516442792147785

<b> As can be seen, when we divide the data set into learning and validation sets, the RMSE increases to 6.494, while the $R^{2}$ and adjusted $R^{2}$ values decrease to 0.49 and 0.485. </b>